// @govtechsg/sgds-web-component/react/components/Table/sgds-table.cjs.js
'use client';
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tslib = require('tslib');
var lit = require('lit');
var decorators_js = require('lit/decorators.js');
var classMap_js = require('lit/directives/class-map.js');
var sgdsElement = require('../../base/sgds-element.cjs.js');
var table = require('./table.cjs.js');

/**
 * @summary The use of a table is to organise a collections of data into readable rows
 */
class SgdsTable extends sgdsElement["default"] {
    constructor() {
        super(...arguments);
        /**
         * Populate row header cells using Arrays
         */
        this.rowHeader = [];
        /**
         * Populate column header cells using Arrays only when <code>headerPosition="vertical"</code> or <code>headerPosition="both"</code>
         */
        this.columnHeader = [];
        /**
         * Populate data cells using Arrays
         */
        this.tableData = [];
        /**
         * Defines the placement of headers in the table (horizontal, vertical, or both)
         */
        this.headerPosition = "horizontal";
        /** @internal */
        this.originalTableData = [];
    }
    connectedCallback() {
        super.connectedCallback();
        this.originalTableData = [...this.tableData];
    }
    _renderTable() {
        if (this.headerPosition === "horizontal") {
            return lit.html `
        <thead>
          <tr>
            ${this.rowHeader.map((header, index) => lit.html ` <th>${header}</th> `)}
          </tr>
        </thead>
        <tbody>
          ${this.tableData.map(row => lit.html `
              <tr>
                ${row.map((cell) => lit.html `<td>${cell}</td>`)}
              </tr>
            `)}
        </tbody>
      `;
        }
        if (this.headerPosition === "both") {
            return lit.html `
        <thead>
          <tr>
            <th></th>
            ${this.rowHeader.map((header, index) => lit.html ` <th>${header}</th> `)}
          </tr>
        </thead>
        <tbody>
          ${this.tableData.map((row, index) => lit.html `
              <tr>
                <th>${this.columnHeader[index]}</th>
                ${row.map((cell) => lit.html `<td>${cell}</td>`)}
              </tr>
            `)}
        </tbody>
      `;
        }
        if (this.headerPosition === "vertical") {
            const flippedTableData = this.tableData[0].map((_, colIndex) => this.tableData.map(row => row[colIndex]));
            return lit.html `
        ${flippedTableData.map((row, index) => lit.html `
            <tr>
              <th>${this.columnHeader[index]}</th>
              ${row.map((cell) => lit.html `<td>${cell}</td>`)}
            </tr>
          `)}
      `;
        }
    }
    render() {
        return lit.html `
      <div
        class=${classMap_js.classMap({
            "table-responsive": this.responsive === "always",
            "table-responsive-sm": this.responsive === "sm",
            "table-responsive-md": this.responsive === "md",
            "table-responsive-lg": this.responsive === "lg",
            "table-responsive-xl": this.responsive === "xl"
        })}
        tabindex="0"
      >
        <table class="table">
          ${this._renderTable()}
        </table>
      </div>
    `;
    }
}
SgdsTable.styles = [...sgdsElement["default"].styles, table["default"]];
tslib.__decorate([
    decorators_js.property({ type: String, reflect: true })
], SgdsTable.prototype, "responsive", void 0);
tslib.__decorate([
    decorators_js.property({ type: Array })
], SgdsTable.prototype, "rowHeader", void 0);
tslib.__decorate([
    decorators_js.property({ type: Array })
], SgdsTable.prototype, "columnHeader", void 0);
tslib.__decorate([
    decorators_js.property({ type: Array })
], SgdsTable.prototype, "tableData", void 0);
tslib.__decorate([
    decorators_js.property({ type: String, reflect: true })
], SgdsTable.prototype, "headerPosition", void 0);
tslib.__decorate([
    decorators_js.state()
], SgdsTable.prototype, "originalTableData", void 0);

exports.SgdsTable = SgdsTable;
exports["default"] = SgdsTable;
