// @govtechsg/sgds-web-component/react/components/Checkbox/sgds-checkbox-group.cjs.js
'use client';
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tslib = require('tslib');
var lit = require('lit');
var decorators_js = require('lit/decorators.js');
var sgdsElement = require('../../base/sgds-element.cjs.js');
var feedback = require('../../styles/feedback.cjs.js');
var formLabel = require('../../styles/form-label.cjs.js');
var checkboxGroup = require('./checkbox-group.cjs.js');
var formHint = require('../../styles/form-hint.cjs.js');

/**
 * @summary CheckboxGroup is the container that group multiple checkboxes under a single question field.
 * It handles the display of validation feedback of its checkboxes children.
 *
 * @slot default - Pass in `sgds-checkbox` into the default slot
 * @slot invalidIcon - The slot for invalid icon
 *
 */
class SgdsCheckboxGroup extends sgdsElement["default"] {
    constructor() {
        super();
        this.hasInvalidCheckbox = false;
        /** The checkbox group's label  */
        this.label = "";
        /**Feedback text for error state when validated */
        this.invalidFeedback = "";
        /** Allows invalidFeedback, invalid styles to be visible. When SgdsCheckboxGroup is used, it overrides the value of hasFeedback on SgdsCheckbox with its own value. */
        this.hasFeedback = false;
        /** The checkbox group's hint text */
        this.hintText = "";
        this.addEventListener("sgds-validity-change", (e) => {
            this.hasInvalidCheckbox = e.detail.invalid;
            this.validationMessage = e.detail.validationMessage;
        });
    }
    _checkInvalidState() {
        this.hasInvalidCheckbox = Array.from(this.checkboxes).some(checkbox => checkbox.invalid);
    }
    /** Overrides hasFeedback from individual SgdsCheckbox  */
    _forwardHasFeedback() {
        Array.from(this.checkboxes).forEach(checkbox => (checkbox.hasFeedback = this.hasFeedback));
    }
    _renderHintText() {
        const hintTextTemplate = lit.html ` <div class="form-text">${this.hintText}</div> `;
        return this.hintText && hintTextTemplate;
    }
    firstUpdated() {
        this._forwardHasFeedback();
    }
    updated() {
        this._checkInvalidState();
    }
    render() {
        return lit.html `
      <fieldset>
        <div class="label-hint-container">
          <label class="form-label">${this.label}</label>
          ${this._renderHintText()}
        </div>
        <div class="checkbox-container">
          <slot></slot>
        </div>
        ${this.hasInvalidCheckbox && this.hasFeedback
            ? lit.html `
              <div class="invalid-feedback-container">
                <slot name="invalidIcon">
                  <svg xmlns="http://www.w3.org/2000/svg" width="20" height="20" viewBox="0 0 20 20" fill="none">
                    <path
                      d="M17.5 10C17.5 14.1421 14.1421 17.5 10 17.5C5.85786 17.5 2.5 14.1421 2.5 10C2.5 5.85786 5.85786 2.5 10 2.5C14.1421 2.5 17.5 5.85786 17.5 10ZM10 6.25C9.49805 6.25 9.10584 6.68339 9.15578 7.18285L9.48461 10.4711C9.51109 10.7359 9.7339 10.9375 10 10.9375C10.2661 10.9375 10.4889 10.7359 10.5154 10.4711L10.8442 7.18285C10.8942 6.68339 10.5019 6.25 10 6.25ZM10.0014 11.875C9.48368 11.875 9.06394 12.2947 9.06394 12.8125C9.06394 13.3303 9.48368 13.75 10.0014 13.75C10.5192 13.75 10.9389 13.3303 10.9389 12.8125C10.9389 12.2947 10.5192 11.875 10.0014 11.875Z"
                      fill="currentColor"
                    />
                  </svg>
                </slot>
                <div id="checkbox-feedback" tabindex="0" class="invalid-feedback">
                  ${this.invalidFeedback ? this.invalidFeedback : this.validationMessage}
                </div>
              </div>
            `
            : lit.nothing}
      </fieldset>
    `;
    }
}
SgdsCheckboxGroup.styles = [...sgdsElement["default"].styles, feedback["default"], formLabel["default"], checkboxGroup["default"], formHint["default"]];
tslib.__decorate([
    decorators_js.queryAssignedElements({ flatten: true })
], SgdsCheckboxGroup.prototype, "checkboxes", void 0);
tslib.__decorate([
    decorators_js.state()
], SgdsCheckboxGroup.prototype, "hasInvalidCheckbox", void 0);
tslib.__decorate([
    decorators_js.state()
], SgdsCheckboxGroup.prototype, "validationMessage", void 0);
tslib.__decorate([
    decorators_js.property({ reflect: true })
], SgdsCheckboxGroup.prototype, "label", void 0);
tslib.__decorate([
    decorators_js.property({ type: String, reflect: true })
], SgdsCheckboxGroup.prototype, "invalidFeedback", void 0);
tslib.__decorate([
    decorators_js.property({ type: Boolean, reflect: true })
], SgdsCheckboxGroup.prototype, "hasFeedback", void 0);
tslib.__decorate([
    decorators_js.property({ reflect: true })
], SgdsCheckboxGroup.prototype, "hintText", void 0);

exports.SgdsCheckboxGroup = SgdsCheckboxGroup;
exports["default"] = SgdsCheckboxGroup;
