// @govtechsg/sgds-web-component/react/components/Badge/sgds-badge.js
'use client';
import { __decorate } from 'tslib';
import { html, nothing } from 'lit';
import { property } from 'lit/decorators.js';
import { classMap } from 'lit/directives/class-map.js';
import SgdsElement from '../../base/sgds-element.js';
import css_248z from './badge.js';
import { watch } from '../../utils/watch.js';
import { SgdsCloseButton } from '../../internals/CloseButton/sgds-close-button.js';

/**
 * @summary Badges can be used to highlight important bits of information such as labels, notifications & status.
 *
 * @slot default - slot for badge
 * @slot icon - The slot for icon to the left of the badge text
 *
 * @event sgds-show - Emitted when the badge appears.
 * @event sgds-hide - Emitted after the badge closes.
 */
class SgdsBadge extends SgdsElement {
    constructor() {
        super(...arguments);
        /** Controls the appearance of the dismissible badge. This prop only applies when dismissible is true  */
        this.show = false;
        /** One or more button variant combinations buttons may be one of a variety of visual variants such as: `info`, `success`, `danger`, `warning`, 'neutral' */
        this.variant = "info";
        /** Manually set the outlined state to false */
        this.outlined = false;
        /** Manually set the dismissible state of the button to `false` */
        this.dismissible = false;
    }
    /** Closes the badge  */
    close() {
        this.show = false;
    }
    /**@internal */
    _handleShowChange() {
        this.show ? this.emit("sgds-show") : this.emit("sgds-hide");
    }
    render() {
        return (this.dismissible && this.show) || !this.dismissible
            ? html `
          <div
            class="  
          ${classMap({
                [`badge-dismissible`]: this.dismissible,
                badge: true,
                outlined: this.outlined
            })}
            "
            aria-hidden=${this.show ? "false" : "true"}
          >
            ${!this.dismissible ? html `<slot name="icon"></slot>` : nothing}
            <span class="badge-label">
              <slot></slot>
            </span>

            ${this.dismissible
                ? html `<sgds-close-button
                  size="sm"
                  aria-label="close the badge"
                  @click=${this.close}
                  variant=${this.outlined ? "dark" : "light"}
                ></sgds-close-button>`
                : nothing}
          </div>
        `
            : nothing;
    }
}
SgdsBadge.styles = [...SgdsElement.styles, css_248z];
/**@internal */
SgdsBadge.dependencies = {
    "sgds-close-button": SgdsCloseButton
};
__decorate([
    property({ type: Boolean, reflect: true })
], SgdsBadge.prototype, "show", void 0);
__decorate([
    property({ reflect: true })
], SgdsBadge.prototype, "variant", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], SgdsBadge.prototype, "outlined", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], SgdsBadge.prototype, "dismissible", void 0);
__decorate([
    watch("show")
], SgdsBadge.prototype, "_handleShowChange", null);

export { SgdsBadge, SgdsBadge as default };
