// @govtechsg/sgds-web-component/react/stepper/index.cjs.js
'use client';
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');
var react = require('@lit/react');
var sgdsStepper = require('../components/Stepper/sgds-stepper.cjs.js');
var ceRegistry = require('../utils/ce-registry.cjs.js');

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n["default"] = e;
  return Object.freeze(n);
}

var React__namespace = /*#__PURE__*/_interopNamespace(React);

ceRegistry.register("sgds-stepper", sgdsStepper.SgdsStepper);
var index = react.createComponent({
    react: React__namespace,
    tagName: "sgds-stepper",
    elementClass: sgdsStepper.SgdsStepper,
    events: {
        onSgdsNextStep: "sgds-next-step",
        onSgdsPreviousStep: "sgds-previous-step",
        onSgdsLastStep: "sgds-last-step",
        onSgdsFirstStep: "sgds-first-step",
        onSgdsArrived: "sgds-arrived",
        onSgdsReset: "sgds-reset"
    }
});

exports["default"] = index;
