// @govtechsg/sgds-web-component/react/quantity-toggle/index.cjs.js
'use client';
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');
var react = require('@lit/react');
var sgdsQuantityToggle = require('../components/QuantityToggle/sgds-quantity-toggle.cjs.js');
var ceRegistry = require('../utils/ce-registry.cjs.js');

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n["default"] = e;
  return Object.freeze(n);
}

var React__namespace = /*#__PURE__*/_interopNamespace(React);

ceRegistry.register("sgds-quantity-toggle", sgdsQuantityToggle.SgdsQuantityToggle);
var index = react.createComponent({
    react: React__namespace,
    tagName: "sgds-quantity-toggle",
    elementClass: sgdsQuantityToggle.SgdsQuantityToggle,
    events: {
        onSgdsChange: "sgds-change",
        onSgdsInput: "sgds-input"
    }
});

exports["default"] = index;
