// @govtechsg/sgds-web-component/react/components/Footer/sgds-footer.js
'use client';
import { __decorate } from 'tslib';
import { html } from 'lit';
import { property } from 'lit/decorators.js';
import SgdsElement from '../../base/sgds-element.js';
import css_248z from './footer.js';

/**
 * @summary The footer contains supporting information for your service at the bottom of your website. All .gov.sg digital services shall contain a Global Footer Bar across all pages. The Global Footer Bar should include the name of the digital service, contact information, a privacy statement and the terms of use.
 *
 * @slot title - The slot for title
 * @slot description - The slot for description
 * @slot items - The slot for the list of footer items
 * @slot social-media - The slot for the list of social media with icons
 *
 */
class SgdsFooter extends SgdsElement {
    constructor() {
        super(...arguments);
        /**
         * 	Sets copyrightLiner of SgdsFooter
         */
        this.copyrightLiner = "Government of Singapore";
        /**
         * 	href link for contacts
         */
        this.contactHref = "#";
        /**
         * 	href link for feedback
         */
        this.feedbackHref = "#";
        /**
         * 	href link for faq
         */
        this.faqHref = "#";
        /**
         * 	href link for privacy statement
         */
        this.privacyHref = "#";
        /**
         * 	href link for terms of use
         */
        this.termsOfUseHref = "#";
    }
    firstUpdated() {
        const socialMediaSlot = this.shadowRoot.querySelector("slot[name='social-media']");
        const footerTitleSlot = this.shadowRoot.querySelector("slot[name='title']");
        const footerDescriptionSlot = this.shadowRoot.querySelector("slot[name='description']");
        const footerItemsSlot = this.shadowRoot.querySelector("slot[name='items']");
        const socialMediaChildNodes = socialMediaSlot.assignedNodes({ flatten: true });
        const footerTitleChildNodes = footerTitleSlot.assignedNodes({ flatten: true });
        const footerDescriptionChildNodes = footerDescriptionSlot.assignedNodes({ flatten: true });
        const footerItemsChildNodes = footerItemsSlot.assignedNodes({ flatten: true });
        if (socialMediaChildNodes.length === 0) {
            const socialMediaContainer = this.shadowRoot.querySelector(".social-media");
            socialMediaContainer.style.display = "none";
        }
        if (footerTitleChildNodes.length === 0 && footerDescriptionChildNodes.length === 0) {
            const footerHeaderContainer = this.shadowRoot.querySelector(".footer-header");
            footerHeaderContainer.style.display = "none";
        }
        if (footerItemsChildNodes.length === 0) {
            const footerItemsContainer = this.shadowRoot.querySelector(".footer-items");
            footerItemsContainer.style.display = "none";
        }
        if (footerTitleChildNodes.length === 0 &&
            footerDescriptionChildNodes.length === 0 &&
            footerItemsChildNodes.length === 0) {
            const footerTopContainer = this.shadowRoot.querySelector(".footer-top");
            footerTopContainer.style.display = "none";
        }
    }
    render() {
        return html `
      <footer class="footer">
        <section class="footer-top">
          <div class="footer-header">
            <slot name="title"></slot>
            <slot name="description"></slot>
          </div>
          <div class="footer-items">
            <slot name="items"></slot>
          </div>
        </section>
        <section class="footer-bottom">
          <div class="social-media">
            <slot name="social-media"></slot>
          </div>
          <div class="footer-mandatory-links">
            <ul>
              <li><a href=${this.contactHref}>Contact</a></li>
              <li><a href=${this.feedbackHref}>Feedback</a></li>
              <li><a href=${this.faqHref}>FAQ</a></li>
              <li>
                <a href="https://tech.gov.sg/report_vulnerability" target="_blank" rel="noopener noreferrer">
                  Report Vulnerability
                </a>
              </li>
              <li><a href=${this.privacyHref}>Privacy Statement</a></li>
              <li><a href=${this.termsOfUseHref}>Terms of use</a></li>
            </ul>
            <div class="footer-copyrights">© ${new Date().getFullYear()}, ${this.copyrightLiner}</div>
          </div>
        </section>
      </footer>
    `;
    }
}
SgdsFooter.styles = [...SgdsElement.styles, css_248z];
__decorate([
    property({ type: String })
], SgdsFooter.prototype, "copyrightLiner", void 0);
__decorate([
    property({ type: String })
], SgdsFooter.prototype, "contactHref", void 0);
__decorate([
    property({ type: String })
], SgdsFooter.prototype, "feedbackHref", void 0);
__decorate([
    property({ type: String })
], SgdsFooter.prototype, "faqHref", void 0);
__decorate([
    property({ type: String })
], SgdsFooter.prototype, "privacyHref", void 0);
__decorate([
    property({ type: String })
], SgdsFooter.prototype, "termsOfUseHref", void 0);

export { SgdsFooter, SgdsFooter as default };
